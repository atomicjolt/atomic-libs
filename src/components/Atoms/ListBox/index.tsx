import React, { useMemo, useRef, useState } from "react";
import type { AriaListBoxProps } from "react-aria";
import { useListState, Node, ListState } from "react-stately";
import {
  mergeProps,
  useFilter,
  useFocusRing,
  useListBox,
  useListBoxSection,
  useOption,
} from "react-aria";
import {
  List,
  ListItem,
  SectionSeperator,
  SectionTitle,
  SubList,
} from "./ListBox.styles";
import Label from "../Label";
import { BaseProps } from "../../../types";
import classNames from "classnames";
import useForwardedRef from "../../../hooks/useForwardedRef";
import SearchInput from "../../Inputs/SearchInput";

export type ListBoxProps<T> = AriaListBoxProps<T> &
  BaseProps & {
    /** Allow the items in the select to be searchable */
    isSearchable?: boolean;
    /** Placeholder for the search input */
    searchPlaceholder?: string;
  };

/** A listbox displays a list of options and allows a user to select one or more of them.
 * Used as the dropdown menu for `ComboBox` and `CustomSelect` */
export function ListBox<T extends object>(props: ListBoxProps<T>) {
  const state = useListState(props);
  return <UnmanagedListBox {...props} state={state} />;
}

export type UnmanagedListBoxProps<T> = ListBoxProps<T> & {
  state: ListState<T>;
};

/** Listbox, but the state is passed in instead of managed internally */
export const UnmanagedListBox = React.forwardRef<
  HTMLUListElement,
  UnmanagedListBoxProps<any>
>((props, ref) => {
  const {
    isSearchable,
    searchPlaceholder,
    state,
    className,
    size = "medium",
  } = props;
  const internalRef = useForwardedRef(ref);
  const { listBoxProps, labelProps } = useListBox(props, state, internalRef);
  const [searchValue, setSearchValue] = useState("");

  const { contains } = useFilter({ sensitivity: "base" });

  const filteredItems = useMemo(
    () =>
      [...state.collection].filter((item) =>
        item.type === "section" ? true : contains(item.textValue, searchValue)
      ),
    [state.collection, searchValue]
  );

  return (
    <>
      <Label {...labelProps}>{props.label}</Label>
      <List
        {...listBoxProps}
        ref={internalRef}
        className={classNames("aje-listbox", className, `is-${size}`)}
      >
        {isSearchable && (
          <SearchInput
            label="Search"
            hideLabel
            value={searchValue}
            onChange={setSearchValue}
            placeholder={searchPlaceholder}
            size={size}
          />
        )}
        {filteredItems.map((item) =>
          item.type === "section" ? (
            <ListBoxSection
              key={item.key}
              section={item}
              state={state}
              filter={(v) => contains(v, searchValue)}
            />
          ) : (
            <ListBoxOption key={item.key} item={item} state={state} />
          )
        )}
      </List>
    </>
  );
});

interface ListBoxSectionProps<T> {
  readonly section: Node<T>;
  readonly state: ListState<T>;
  readonly filter: (text: string) => boolean;
}

function ListBoxSection<T>(props: ListBoxSectionProps<T>) {
  const { section, state, filter } = props;
  const { itemProps, headingProps, groupProps } = useListBoxSection({
    heading: section.rendered,
    "aria-label": section["aria-label"],
  });

  const children = [...section.childNodes].filter((item) =>
    filter(item.textValue)
  );

  if (children.length === 0) {
    return null;
  }

  // If the section is not the first, add a separator element to provide visual separation.
  // The heading is rendered inside an <li> element, which contains
  // a <ul> with the child items.
  return (
    <>
      {section.key !== state.collection.getFirstKey() && (
        <SectionSeperator role="presentation" />
      )}
      <li {...itemProps}>
        {section.rendered && (
          <SectionTitle {...headingProps}>{section.rendered}</SectionTitle>
        )}
        <SubList {...groupProps}>
          {children.map((node) => (
            <ListBoxOption key={node.key} item={node} state={state} />
          ))}
        </SubList>
      </li>
    </>
  );
}

interface ListBoxOptionProps<T> {
  item: Node<T>;
  state: ListState<T>;
}

function ListBoxOption<T>(props: ListBoxOptionProps<T>) {
  const { item, state } = props;
  const ref = useRef(null);
  const { optionProps } = useOption({ key: item.key }, state, ref);

  // Determine whether we should show a keyboard
  // focus ring for accessibility
  const { focusProps } = useFocusRing();

  return (
    <ListItem
      {...mergeProps(optionProps, focusProps)}
      ref={ref}
      data-focus-visible={item.key === state.selectionManager.focusedKey}
    >
      {item.rendered}
    </ListItem>
  );
}
