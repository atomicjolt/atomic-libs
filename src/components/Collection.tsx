import { Node } from "react-stately";
import {
  ItemProps as RSItemProps,
  SectionProps as RSSectionProps,
  Item as RSItem,
  Section as RSSection,
} from "react-stately";
import {
  createBranchComponent,
  createLeafComponent,
} from "@react-aria/collections";
import { createContext, useContext } from "react";
import { copyStaticProperties } from "@utils/clone";
import { Key, RenderBaseProps, RenderStyleProps } from "../types";

export interface ElementItemProps<T, R extends object = object>
  extends Omit<RSItemProps<T>, "children">,
    RenderBaseProps<R> {
  id?: Key;
}

interface ItemContextValue {
  render: (
    props: ElementItemProps<any>,
    ref: React.ForwardedRef<HTMLElement>,
    item: Node<any>
  ) => React.ReactElement;
}

export const ItemContext = createContext<ItemContextValue | null>(null);

/** An Item in a Collection component
 * @example
 * <CustomSelect label="Select a fruit">
 *  <Item id="apple">Apple</Item>
 *  <Item id="orange">Orange</Item>
 *  <Item id="banana">Banana</Item>
 * </CustomSelect>
 */
export const Item = createLeafComponent(
  "item",
  (
    props: ElementItemProps<any>,
    ref: React.ForwardedRef<HTMLElement>,
    item: Node<object>
  ) => {
    const { render } = useContext(ItemContext)!;
    return render(props, ref, item);
  }
);

copyStaticProperties(RSItem, Item);

export interface ElementSectionProps<T, R extends object = object>
  extends RSSectionProps<T>,
    RenderStyleProps<R> {
  id?: Key;
}

interface SectionContextValue {
  render: (
    props: ElementSectionProps<any>,
    ref: React.ForwardedRef<HTMLElement>,
    section: Node<any>
  ) => React.ReactElement;
}

export const SectionContext = createContext<SectionContextValue | null>(null);

/** A Section / grouping of items in a collection component
 * @example
 * <ListBox>
 *  <Section title="Fruits">
 *    <Item id="apple">Apple</Item>
 *    <Item id="orange">Orange</Item>
 *    <Item id="banana">Banana</Item>
 *  </Section>
 *  <Section title="Vegetables">
 *    <Item id="carrot">Carrot</Item>
 *    <Item id="potato">Potato</Item>
 *    <Item id="onion">Onion</Item>
 *  </Section>
 * </ListBox>
 */
export const Section = createBranchComponent(
  "section",
  (
    props: ElementSectionProps<any>,
    ref: React.ForwardedRef<HTMLElement>,
    section: Node<object>
  ) => {
    const { render } = useContext(SectionContext)!;
    return render(props, ref, section);
  }
);

copyStaticProperties(RSSection, Section);

export type {
  ElementItemProps as ItemProps,
  ElementSectionProps as SectionProps,
};
