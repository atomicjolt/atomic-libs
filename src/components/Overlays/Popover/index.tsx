import React, { useCallback, useContext, useMemo, useState } from "react";
import {
  AriaPopoverProps,
  DismissButton,
  Overlay,
  usePopover,
} from "@react-aria/overlays";
import { OverlayTriggerProps, useOverlayTriggerState } from "react-stately";
import { mergeProps } from "@react-aria/utils";

import { HasClassName, HasVariant } from "../../../types";
import { useRenderProps } from "@hooks";
import { useContextProps } from "@hooks/useContextProps";
import { useResizeObserver } from "@hooks/useResizeObserver";
import { useForwardedRef } from "@hooks/useForwardedRef";
import { PopoverUnderlay, PopoverContent } from "./Popover.styles";
import { PopoverContext } from "./context";
import { OverlayTriggerStateContext } from "../OverlayTrigger/context";
import { invertPlacementAxis } from "@utils/placement";

export interface PopoverProps
  extends Omit<AriaPopoverProps, "popoverRef" | "triggerRef">,
    OverlayTriggerProps,
    HasClassName,
    HasVariant<"listbox" | "menu" | "datepicker"> {
  children: React.ReactNode;
  /** A ref to the element that triggers the popover to appear. Not necessary when used with trigger wrapper */
  triggerRef?: React.RefObject<HTMLElement>;
  id?: string;
}

/** A popover is an overlay element positioned relative to a target. */
export const Popover = React.forwardRef<HTMLDivElement, PopoverProps>(
  (props: PopoverProps, ref) => {
    const contextProps = useContextProps(PopoverContext, props);
    const {
      children,
      offset = 8,
      triggerRef,
      variant,
      placement = "bottom",
      ...rest
    } = contextProps;
    const contextState = useContext(OverlayTriggerStateContext);
    const localState = useOverlayTriggerState(props);
    const state = contextState ?? localState;

    const internalRef = useForwardedRef(ref);

    const { popoverProps, underlayProps } = usePopover(
      {
        ...rest,
        placement,
        offset,
        triggerRef: triggerRef!,
        popoverRef: internalRef,
      },
      state
    );

    const [triggerWidth, setTriggerWidth] = useState(
      `${triggerRef?.current?.offsetWidth || 0}px`
    );

    useResizeObserver(
      triggerRef?.current,
      useCallback(
        () => setTriggerWidth(`${triggerRef?.current?.offsetWidth || 0}px`),
        []
      )
    );

    const renderProps = useRenderProps({
      componentClassName: "aje-popover",
      ...contextProps,
    });

    const transformOrigin = invertPlacementAxis(placement);

    if (!state.isOpen) {
      return null;
    }

    return (
      <Overlay>
        <PopoverUnderlay {...underlayProps} className="aje-popover-underlay" />
        <PopoverContent
          ref={internalRef}
          id={rest.id}
          {...mergeProps(popoverProps, renderProps)}
          style={{
            ...popoverProps.style,
            ...renderProps.style,
            // @ts-ignore
            "--trigger-width": triggerWidth,
          }}
          $transformOrigin={transformOrigin}
        >
          <DismissButton onDismiss={state.close} />
          {children}
          <DismissButton onDismiss={state.close} />
        </PopoverContent>
      </Overlay>
    );
  }
);
