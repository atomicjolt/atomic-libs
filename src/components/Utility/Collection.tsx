import {
  Item as ReactStatelyItem,
  Section as ReactStatelySection,
  ItemProps,
  SectionProps,
} from "react-stately";

// We re-define the react-stately Item and Section components
// Because for some reason storybook doesn't like to pickup the
// correct names for the reac-stately components.

/** An Item in a Collection component
 * @example
 * <CustomSelect label="Select a fruit">
 *  <Item key="apple">Apple</Item>
 *  <Item key="orange">Orange</Item>
 *  <Item key="banana">Banana</Item>
 * </CustomSelect>
 */
export function Item<T>(props: ItemProps<T>) {
  return null;
}

// React Stately binds some static properties to the Item component that we need to copy over
for (const key in ReactStatelyItem) {
  // @ts-ignore
  Item[key] = ReactStatelyItem[key];
}

/** A Section / grouping of items in a collection component
 * @example
 * <ListBox>
 *  <Section title="Fruits">
 *    <Item key="apple">Apple</Item>
 *    <Item key="orange">Orange</Item>
 *    <Item key="banana">Banana</Item>
 *  </Section>
 *  <Section title="Vegetables">
 *    <Item key="carrot">Carrot</Item>
 *    <Item key="potato">Potato</Item>
 *    <Item key="onion">Onion</Item>
 *  </Section>
 * </ListBox>
 */
export function Section<T>(props: SectionProps<T>) {
  return null;
}

for (const key in ReactStatelySection) {
  // @ts-ignore
  Section[key] = ReactStatelySection[key];
}
