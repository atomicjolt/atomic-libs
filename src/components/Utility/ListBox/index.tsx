import React, { useRef } from "react";
import type { AriaListBoxProps } from "react-aria";
import { useListState, Node, ListState } from "react-stately";
import {
  mergeProps,
  useFocusRing,
  useListBox,
  useListBoxSection,
  useOption,
} from "react-aria";
import {
  List,
  ListItem,
  SectionSeperator,
  SectionTitle,
  SubList,
} from "./ListBox.styles";
import Label from "../Label";
import { BaseProps } from "../../../types";
import classNames from "classnames";
import useForwardedRef from "../../../hooks/useForwardedRef";

export type ListBoxProps<T> = AriaListBoxProps<T> & BaseProps;

/** A listbox displays a list of options and allows a user to select one or more of them.
 * Used as the dropdown menu for `ComboBox` and `CustomSelect` */
export function ListBox<T extends object>(props: ListBoxProps<T>) {
  const state = useListState(props);
  return <UnmanagedListBox {...props} state={state} />;
}

export type UnmanagedListBoxProps<T> = ListBoxProps<T> & {
  state: ListState<T>;
};

/** Listbox, but the state is passed in instead of managed internally */
export const UnmanagedListBox = React.forwardRef<
  HTMLUListElement,
  UnmanagedListBoxProps<any>
>((props, ref) => {
  const { state, className, size = "medium" } = props;
  const internalRef = useForwardedRef(ref);
  const { listBoxProps, labelProps } = useListBox(props, state, internalRef);

  return (
    <>
      <Label {...labelProps}>{props.label}</Label>
      <List
        {...listBoxProps}
        ref={internalRef}
        className={classNames("aje-listbox", className, `is-${size}`)}
      >
        <input role="searchbox" />
        {[...state.collection].map((item) =>
          item.type === "section" ? (
            <ListBoxSection key={item.key} section={item} state={state} />
          ) : (
            <ListBoxOption key={item.key} item={item} state={state} />
          )
        )}
      </List>
    </>
  );
});

interface ListBoxSectionProps<T> {
  section: Node<T>;
  state: ListState<T>;
}

function ListBoxSection<T>(props: ListBoxSectionProps<T>) {
  const { section, state } = props;
  let { itemProps, headingProps, groupProps } = useListBoxSection({
    heading: section.rendered,
    "aria-label": section["aria-label"],
  });

  // If the section is not the first, add a separator element to provide visual separation.
  // The heading is rendered inside an <li> element, which contains
  // a <ul> with the child items.
  return (
    <>
      {section.key !== state.collection.getFirstKey() && (
        <SectionSeperator role="presentation" />
      )}
      <li {...itemProps}>
        {section.rendered && (
          <SectionTitle {...headingProps}>{section.rendered}</SectionTitle>
        )}
        <SubList {...groupProps}>
          {[...section.childNodes].map((node) => (
            <ListBoxOption key={node.key} item={node} state={state} />
          ))}
        </SubList>
      </li>
    </>
  );
}

interface ListBoxOptionProps<T> {
  item: Node<T>;
  state: ListState<T>;
}

function ListBoxOption<T>(props: ListBoxOptionProps<T>) {
  const { item, state } = props;
  let ref = useRef(null);
  let { optionProps } = useOption({ key: item.key }, state, ref);

  // Determine whether we should show a keyboard
  // focus ring for accessibility
  let { focusProps } = useFocusRing();

  return (
    <ListItem
      {...mergeProps(optionProps, focusProps)}
      ref={ref}
      data-focus-visible={item.key === state.selectionManager.focusedKey}
    >
      {item.rendered}
    </ListItem>
  );
}
