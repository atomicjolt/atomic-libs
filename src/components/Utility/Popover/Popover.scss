@use "../../mixins.scss";

.aje-popover__wrapper {
  position: relative;
  display: inline-block;
}

.aje-popover {
  --popover-animation: transform 200ms ease, opacity 200ms;
  --popover-disatnce: 5px;

  visibility: hidden;
  position: absolute;
  width: var(--popover-width, auto);
  z-index: 100;
  opacity: 0;
  transform: scaleY(0.9);
  transform-origin: top;
  transition: var(--popover-animation);

  &-bottom-left {
    margin-top: var(--popover-disatnce);
    top: 100%;
    left: 0;
  }

  &-bottom-right {
    margin-top: var(--popover-disatnce);
    top: 100%;
    right: 0;
  }

  &-top-left {
    margin-bottom: var(--popover-disatnce);
    bottom: 100%;
    left: 0;
  }

  &-top-right {
    margin-bottom: var(--popover-disatnce);
    bottom: 100%;
    right: 0;
  }

  &-left {
    margin-right: var(--popover-disatnce);
    right: 100%;
    top: 0;
  }

  &-right {
    margin-left: var(--popover-disatnce);
    left: 100%;
    top: 0;
  }

  &.is-small {
    --popover-width: 100px;
  }
  &.is-medium {
    --popover-width: 200px;
  }
  &.is-large {
    --popover-width: 300px;
  }
  &.is-full {
    --popover-width: 100%;
  }
  &.is-auto {
    --popover-width: max-content;
  }

  &.is-visible {
    visibility: visible;
    opacity: 1;
    transform: scaleY(1);
  }

  .dropdown {
    background-color: lightblue;
    padding: 10px;
    @include mixins.regular;
    font-size: 20px;
  }
}
