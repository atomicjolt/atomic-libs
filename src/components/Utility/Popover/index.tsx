import React from "react";
import cn from "classnames";
import { Sizes } from "../../../types";

export type PopoverPosition =
  | "bottom-right"
  | "bottom-left"
  | "top-right"
  | "top-left"
  | "left"
  | "right";

export interface PopoverProps {
  children: React.ReactNode;
  size?: Sizes;
  show?: boolean;
  position?: PopoverPosition;
}

export default function Popover(props: PopoverProps) {
  const { children, show, size, position = "bottom-left" } = props;

  return (
    <div
      className={cn("aje-popover", `is-${size}`, `aje-popover-${position}`, {
        "is-visible": show,
      })}
    >
      {children}
    </div>
  );
}

interface PopoverWrapperProps {
  children: React.ReactNode;
}

export function PopoverWrapper(props: PopoverWrapperProps) {
  const { children } = props;

  return <div className={cn("aje-popover__wrapper")}>{children}</div>;
}
