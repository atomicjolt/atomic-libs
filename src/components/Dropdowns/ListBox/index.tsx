import React, { useMemo, useRef, useState } from "react";
import type { AriaListBoxProps } from "react-aria";
import { useListState, Node, ListState } from "react-stately";
import {
  mergeProps,
  useFilter,
  useFocusRing,
  useListBox,
  useListBoxSection,
  useOption,
} from "react-aria";
import classNames from "classnames";
import { BaseProps } from "../../../types";
import useForwardedRef from "../../../hooks/useForwardedRef";
import SearchInput from "../../Inputs/SearchInput";
import MaterialIcon from "../../Icons/MaterialIcon";
import { Label } from "../../Fields";
import { Divider } from "@components/Layout/Divider";
import { List, ListItem, SectionTitle, SubList } from "./ListBox.styles";

export type ListBoxProps<T> = AriaListBoxProps<T> &
  BaseProps & {
    /** Allow the items in the select to be searchable */
    isSearchable?: boolean;
    /** Placeholder for the search input */
    searchPlaceholder?: string;
    /** Whether to show a checkmark next to selected items */
    showCheckmark?: boolean;
  };

/** A listbox displays a list of options and allows a user to select one or more of them.
 * Used as the dropdown menu for `ComboBox` and `CustomSelect` */
export function ListBox<T extends object>(props: ListBoxProps<T>) {
  const state = useListState(props);
  return <UnmanagedListBox {...props} state={state} />;
}

export type UnmanagedListBoxProps<T> = Omit<ListBoxProps<T>, "children"> & {
  state: ListState<T>;
};

/** Listbox, but the state is passed in instead of managed internally */
export const UnmanagedListBox = React.forwardRef<
  HTMLUListElement,
  UnmanagedListBoxProps<any>
>((props, ref) => {
  const {
    isSearchable,
    searchPlaceholder = "Search",
    state,
    className,
    size = "medium",
    showCheckmark = false,
  } = props;
  const internalRef = useForwardedRef(ref);
  const { listBoxProps, labelProps } = useListBox(props, state, internalRef);
  const [searchValue, setSearchValue] = useState("");
  const [isSearching, setIsSearching] = useState(false);

  const { contains } = useFilter({ sensitivity: "base" });

  const filteredItems = useMemo(
    () =>
      [...state.collection].filter((item) =>
        item.type === "section" ? true : contains(item.textValue, searchValue)
      ),
    [state.collection, searchValue]
  );

  if (isSearching) {
    // Prevent the listbox from handling keyboard events while the search input is focused
    // Stops it from taking focus away from the search input
    delete listBoxProps.onKeyDownCapture;
  }

  return (
    <>
      <Label {...labelProps}>{props.label}</Label>
      <List
        {...listBoxProps}
        ref={internalRef}
        className={classNames("aje-listbox", className, `is-${size}`)}
      >
        {isSearchable && (
          <SearchInput
            aria-label="Search"
            value={searchValue}
            onChange={setSearchValue}
            placeholder={searchPlaceholder}
            size={size}
            onFocus={() => setIsSearching(true)}
            onBlur={() => setIsSearching(false)}
          />
        )}
        {filteredItems.map((item) =>
          item.type === "section" ? (
            <ListBoxSection
              key={item.key}
              section={item}
              state={state}
              filter={(v) => contains(v, searchValue)}
              showCheckmark={showCheckmark}
            />
          ) : (
            <ListBoxOption
              key={item.key}
              item={item}
              state={state}
              showCheckmark={showCheckmark}
            />
          )
        )}
      </List>
    </>
  );
});

interface ListBoxSectionProps<T> {
  readonly section: Node<T>;
  readonly state: ListState<T>;
  readonly filter: (text: string) => boolean;
  readonly showCheckmark?: boolean;
}

function ListBoxSection<T>(props: ListBoxSectionProps<T>) {
  const { section, state, filter, showCheckmark = false } = props;
  const { itemProps, headingProps, groupProps } = useListBoxSection({
    heading: section.rendered,
    "aria-label": section["aria-label"],
  });

  const children = [...section.childNodes].filter((item) =>
    filter(item.textValue)
  );

  if (children.length === 0) {
    return null;
  }

  // If the section is not the first, add a separator element to provide visual separation.
  // The heading is rendered inside an <li> element, which contains
  // a <ul> with the child items.
  return (
    <>
      {section.key !== state.collection.getFirstKey() && <Divider as="li" />}
      <li {...itemProps}>
        {section.rendered && (
          <SectionTitle {...headingProps}>{section.rendered}</SectionTitle>
        )}
        <SubList {...groupProps}>
          {children.map((node) => (
            <ListBoxOption
              key={node.key}
              item={node}
              state={state}
              showCheckmark={showCheckmark}
            />
          ))}
        </SubList>
      </li>
    </>
  );
}

interface ListBoxOptionProps<T> {
  item: Node<T>;
  state: ListState<T>;
  showCheckmark?: boolean;
}

function ListBoxOption<T>(props: ListBoxOptionProps<T>) {
  const { item, state, showCheckmark } = props;
  const ref = useRef(null);
  const { optionProps } = useOption({ key: item.key }, state, ref);

  // Determine whether we should show a keyboard
  // focus ring for accessibility
  const { focusProps } = useFocusRing();

  return (
    <ListItem
      {...mergeProps(optionProps, focusProps)}
      ref={ref}
      data-focus-visible={item.key === state.selectionManager.focusedKey}
    >
      {item.rendered}
      {showCheckmark && <MaterialIcon icon="check" />}
    </ListItem>
  );
}
