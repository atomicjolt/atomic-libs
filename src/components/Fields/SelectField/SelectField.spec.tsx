import { describe, test, expect } from "vitest";
import { render } from "@testing-library/react";
import { Popover } from "@components/Overlays/Popover";
import { ListBox } from "@components/Dropdowns/ListBox";
import { Item } from "@components/Collection";
import { SelectField } from ".";

describe("SelectField", () => {
  describe("matches snapshots", () => {
    test("when closed", () => {
      const result = render(
        <SelectField>
          <SelectField.Button>
            <SelectField.Value />
          </SelectField.Button>
          <Popover>
            <ListBox>
              <Item>Item 1</Item>
              <Item>Item 2</Item>
              <Item>Item 3</Item>
            </ListBox>
          </Popover>
        </SelectField>
      );
      expect(result.asFragment()).toMatchSnapshot();
    });

    test("when open", () => {
      const result = render(
        <SelectField isOpen>
          <SelectField.Button>
            <SelectField.Value />
          </SelectField.Button>
          <Popover>
            <ListBox>
              <Item>Item 1</Item>
              <Item>Item 2</Item>
              <Item>Item 3</Item>
            </ListBox>
          </Popover>
        </SelectField>
      );
      expect(result.asFragment()).toMatchSnapshot();
    });
  });
});
