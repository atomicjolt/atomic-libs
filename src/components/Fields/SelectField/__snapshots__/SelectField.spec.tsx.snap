// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelectField > matches snapshots > when closed 1`] = `
<DocumentFragment>
  <div
    class="sc-feUYzb hzHmVu aje-select"
  >
    <div
      aria-hidden="true"
      data-a11y-ignore="aria-hidden-focus"
      data-react-aria-prevent-focus="true"
      data-testid="hidden-select-container"
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        style="font-size: 16px;"
        tabindex="-1"
        type="text"
      />
      <label>
        <select
          tabindex="-1"
        >
          <option />
          <option
            value="react-aria-1"
          >
            Item 1
          </option>
          <option
            value="react-aria-2"
          >
            Item 2
          </option>
          <option
            value="react-aria-3"
          >
            Item 3
          </option>
        </select>
      </label>
    </div>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="react-aria-:r7: react-aria-:r3:"
      class="sc-dLNtp sc-hzhKNl hkAcSC hhPoUm aje-btn sc-fHjqbK hMkkGO aje-btn--dropdown is-medium"
      id="react-aria-:r2:"
      type="button"
    >
      <span
        class="sc-fUnNpA iyHtdr"
        id="react-aria-:r7:"
      />
      <i
        aria-hidden="true"
        class="sc-cPiJYC wZRQV aje-icon material-icons is-medium"
      >
        arrow_drop_down
      </i>
    </button>
  </div>
</DocumentFragment>
`;

exports[`SelectField > matches snapshots > when open 1`] = `
<DocumentFragment>
  <div
    class="sc-feUYzb hzHmVu aje-select"
  >
    <div
      aria-hidden="true"
      data-a11y-ignore="aria-hidden-focus"
      data-react-aria-prevent-focus="true"
      data-testid="hidden-select-container"
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        style="font-size: 16px;"
        tabindex="-1"
        type="text"
      />
      <label>
        <select
          tabindex="-1"
        >
          <option />
          <option
            value="react-aria-1"
          >
            Item 1
          </option>
          <option
            value="react-aria-2"
          >
            Item 2
          </option>
          <option
            value="react-aria-3"
          >
            Item 3
          </option>
        </select>
      </label>
    </div>
    <button
      aria-controls="react-aria-:r9:"
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-labelledby="react-aria-:rf: react-aria-:rb:"
      class="sc-dLNtp sc-hzhKNl hkAcSC hhPoUm aje-btn sc-fHjqbK hMkkGO aje-btn--dropdown is-medium"
      data-pressed="true"
      id="react-aria-:ra:"
      type="button"
    >
      <span
        class="sc-fUnNpA iyHtdr"
        id="react-aria-:rf:"
      />
      <i
        aria-hidden="true"
        class="sc-cPiJYC wZRQV aje-icon material-icons is-medium"
      >
        arrow_drop_down
      </i>
    </button>
  </div>
</DocumentFragment>
`;
